internal/oci/runtime_vm.go

```diff
@@ -109,13 +111,32 @@ func (r *runtimeVM) CreateContainer(c *Container, cgroupParent string) (err erro
 	containerIO.AddOutput("logfile", f, f)
 	containerIO.Pipe()
 
+	r.Lock()
 	r.ctrs[c.ID()] = containerInfo{
 		cio: containerIO,
 	}
+	r.Unlock()
 
 	defer func() {
 		if err != nil {
+			logrus.WithError(err).Warnf("Cleaning up container %v", c.ID())
```

Comment:
is all of the code here in delete? why not add a fatal bool that returns in an error case to a deleteContainer helper function. Or, you can take the step by step approach as others (the moment you provision a resource, defer its removal). or leave it as is :smiley:

Reply:
(the deleteContainer helper could also be non-locking)

Reply:
Yep, all of the code here is in DeleteContainer.

The very first version of patch I added a `cleanupContainer()` helper, which would take a `bool` to meaning force, with some lines of documentation saying that locking was up to the caller, but I must admit it could be more confusing than the approach taken in case someone decides to use it later on and, for some reason, doesn't carefully read the comments.

Reply:
in general I'd prefer to have it in one place. I'd personally like to assume anyone adding a call to a function should have at least read its comment. It add maintenance overhead (and duplicated code) to have this in two places. what if there's a new step to cleaning up a container? more worth it to me to have it in one place

Reply:
Ack!